src/ert/run_models/base_run_model.py

```diff
@@ -495,6 +496,60 @@ def send_snapshot_event(self, event: CloudEvent) -> None:
                 )
             )
 
+    async def run_monitor(self, ee_config: EvaluatorServerConfig) -> bool:
+        try:
+            event_logger.debug("connecting to new monitor...")
+            async with MonitorAsync(ee_config.get_connection_info()) as monitor:
+                event_logger.debug("connected")
+                async for event in monitor.track():
+                    if event["type"] in (
+                        EVTYPE_EE_SNAPSHOT,
+                        EVTYPE_EE_SNAPSHOT_UPDATE,
+                    ):
+                        self.send_snapshot_event(event)
+                        if event.data.get(STATUS) in [
+                            ENSEMBLE_STATE_STOPPED,
+                            ENSEMBLE_STATE_FAILED,
+                        ]:
+                            event_logger.debug(
+                                "observed evaluation stopped event, signal done"
+                            )
+                            await monitor.signal_done()
+                            # should_exit = True
+                        if event.data.get(STATUS) == ENSEMBLE_STATE_CANCELLED:
+                            event_logger.debug(
+                                "observed evaluation cancelled event, exit drainer"
+                            )
+                            # Allow track() to emit an EndEvent.
+                            return False
+                    elif event["type"] == EVTYPE_EE_TERMINATED:
+                        event_logger.debug("got terminator event")
+
+                    if not self._end_queue.empty():
+                        event_logger.debug("Run model canceled - during evaluation")
+                        self._end_queue.get()
+                        await monitor.signal_cancel()
+                        event_logger.debug(
+                            "Run model canceled - during evaluation - cancel sent"
```

Comment:
Can we put most of this, if not all of it, in a separate method? Maybe something like `self._handle_monitor_event(event)` would work.